db/snapshot-ctl.cc

```diff
@@ -144,82 +146,58 @@ future<> snapshot_ctl::clear_snapshot(sstring tag, std::vector<sstring> keyspace
-                return cf_pair.second->get_snapshot_details().then([uuid = cf_pair.first, local_snapshots] (auto map) {
-                    for (auto&& snap_map: map) {
-                        auto [it, ignored] = local_snapshots->try_emplace(snap_map.first);
-                        it->second.emplace(uuid, snap_map.second);
-                    }
-                    return make_ready_future<>();
-                });
-            }).then([local_snapshots] {
-                return make_ready_future<snapshot_map>(std::move(*local_snapshots));
-            });
-        }).then([this] (snapshot_map&& map) {
-            std::unordered_map<sstring, std::vector<snapshot_ctl::snapshot_details>> result;
-            for (auto&& pair: map) {
-                std::vector<snapshot_ctl::snapshot_details> details;
-
-                for (auto&& snap_map: pair.second) {
-                    auto& cf = _db.local().find_column_family(snap_map.first);
-                    details.push_back({ snap_map.second.live, snap_map.second.total, cf.schema()->cf_name(), cf.schema()->ks_name() });
-                }
-                result.emplace(pair.first, std::move(details));
-            }
-
-            return make_ready_future<std::unordered_map<sstring, std::vector<snapshot_ctl::snapshot_details>>>(std::move(result));
+    co_return co_await run_snapshot_list_operation([this] () -> future<snapshot_map> {
+        snapshot_acc collected = co_await _db.map_reduce(snapshot_reducer(), [] (replica::database& db) -> future<std::vector<replica::database::snapshot_details_result>> {
+            co_return co_await db.get_snapshot_details();
         });
+
+        snapshot_map result;
+        for (auto&& r : collected) {
+            result[r.snapshot_name].emplace_back(std::move(r.details));
+
+            co_await coroutine::maybe_yield();
+        }
+
+        co_return result;
     });
 }
 
 future<int64_t> snapshot_ctl::true_snapshots_size() {
-    return run_snapshot_list_operation([this] {
-        return _db.map_reduce(adder<int64_t>(), [] (replica::database& db) {
-            return do_with(int64_t(0), [&db] (auto& local_total) {
-                return parallel_for_each(db.get_column_families(), [&local_total] (auto& cf_pair) {
-                    return cf_pair.second->get_snapshot_details().then([&local_total] (auto map) {
-                        for (auto&& snap_map: map) {
-                            local_total += snap_map.second.live;
-                        }
-                        return make_ready_future<>();
-                     });
-                }).then([&local_total] {
-                    return make_ready_future<int64_t>(local_total);
-                });
-            });
+    co_return co_await run_snapshot_list_operation([this] () -> future<int64_t> {
+        co_return co_await _db.map_reduce(adder<int64_t>(), [] (replica::database& db) -> future<int64_t> {
```

Comment:
One thing I don't understand (and wanted to test before merging the previous PR, but didn't get to it) is why we have to reduce over all shards? It seems that `db.get_snapshot_details()` should already return all snapshots already on one shard (any shard), since it scans all the directories.

That said, I also don't understand why the previous version had to reduce over shards.

Reply:
I agree, the snapshot directory is shared.
Will fix.

Reply:
Does it mean that the previous version was buggy as well?

Reply:
Yes, `snapshot_ctl::true_snapshots_size` is inflated in a factor of `smp::count`.
We may want to fix that independently and backport the fix (with no coroutines).
Let's open an issue and I'll extract the patch fixing it so it can be backported.

Reply:
https://github.com/scylladb/scylla/issues/9897